src/CMakeLists.txt

```diff
@@ -3,28 +3,46 @@
 cmake_minimum_required(VERSION 3.8)
 
 # set compile options
-set(COMMON_C_FLAGS "-Wall -fsigned-char")
-set(COMMON_CXX_FLAGS "-fno-rtti -ffast-math")
+set(COMMON_CXX_FLAGS "-fno-rtti -ffast-math -fexceptions -Wall -fsigned-char")
 
 if(CMAKE_BUILD_TYPE MATCHES RELEASE)
     message(STATUS "Release build detected, adding release flags")
-    set(COMMON_C_FLAGS "${CMAKE_CXX_FLAGS} -O3 -fomit-frame-pointer")
+    set(COMMON_CXX_FLAGS "${CMAKE_CXX_FLAGS} -O2 -fomit-frame-pointer -fstack-protector-all -fpic -fpie -D_FORTIFY_SOURCE=2")
+    if(${CMAKE_SYSTEM_NAME} MATCHES "Linux")
+        set(COMMON_CXX_FLAGS "${ADDITIONAL_BUILD_FLAGS} -D_GLIBCXX_ASSERTIONS")
+    endif()
+    # TODO: "-Wl,-pie"
+    set(CMAKE_EXE_LINKER_FLAGS "-Wl,-z,relro -Wl,-z,now -Wl,-z,defs -pie")
 endif()
 
 option(PEDANTIC "Enable pedantic mode")
 if(PEDANTIC)
     message(STATUS "Enabling pedantic mode")
     if(CMAKE_CXX_COMPILER_ID MATCHES GNU)
-        set(COMMON_C_FLAGS "${ADDITIONAL_BUILD_FLAGS} -Wextra")
+        set(COMMON_CXX_FLAGS "${ADDITIONAL_BUILD_FLAGS} -Wextra")
     elseif(CMAKE_CXX_COMPILER_ID MATCHES Clang)
-        set(COMMON_C_FLAGS "${ADDITIONAL_BUILD_FLAGS} -Wpedantic")
+        set(COMMON_CXX_FLAGS "${ADDITIONAL_BUILD_FLAGS} -Wpedantic")
     else()
         message(FATAL_ERROR "Pedantic mode requested, but no configuration available for compiler identification '${CMAKE_CXX_COMPILER_ID}'")
     endif()
 endif()
 
-set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${COMMON_C_FLAGS} ${COMMON_CXX_FLAGS}")
-set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${COMMON_C_FLAGS}")
+option(HARDENED "Enable hardening options")
+if(HARDENED)
+    message(STATUS "Enabling hardening options")
+    if(CMAKE_CXX_COMPILER_ID MATCHES GNU)
+	set(COMMON_CXX_FLAGS "${ADDITIONAL_BUILD_FLAGS} -fstack-clash-protection")
+    elseif(CMAKE_CXX_COMPILER_ID MATCHES Clang AND CMAKE_CXX_COMPILER_VERSION GREATER 10)
+        set(COMMON_CXX_FLAGS "${ADDITIONAL_BUILD_FLAGS} -fstack-clash-protection -fsanitize=safe-stack -mretpoline -mspeculative-load-hardening -mlvi-cfi")
+	set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -fsanitize=safe-stack -mspeculative-load-hardening -mlvi-cfi")
+    elseif(CMAKE_CXX_COMPILER_VERSION VERSION_LESS 11)
```

Comment:
That one only really works for clang, too, but the check is arguably incomplete. Consider adding one `if` checking for clang, then do the version check stuff within there.